README.md
# growatt-rs232-reader

This reader allows some Growatt Inverters to publish data on a REST endpoint and optionally MQTT topic.

Example output of ```http://127.0.0.1:5701/status```:
```json
{
  "VoltagePV1": 335.8,
  "VoltagePV2": 0,
  "VoltageBus": 380.6,
  "VoltageGrid": 225.7,                   *
  "TotalProduction": 3822.6,              *
  "DayProduction": 0.7,                   *
  "Frequency": 49.99,                     *
  "Power": 798.4,                         *
  "Temperature": 26.9,                    *
  "OperationHours": 1891.338,             *
  "Status": "Normal",
  "FaultCode": 0,
  "Timestamp": "2018-12-09T13:15:54.363021599+01:00"
}
```
Starred fields are optional and won't be included outside status 'Normal'.
MQTT messages will only be send if a value changes (no additional information will be send).

Additional information can be retrieved using: ```curl http://localhost:5701/info```:

```json
{
  "Reader":"Reading since 14:11:18",
  "Interpreter":"Polling since 14:11:18",
  "Publisher":"Normal",
  "Init":"OK"
}
```

Above is a perfectly legal state as long as the times are within 10 minutes of the current time. Note that startup takes several seconds.

## Usage

Usage: ./growatt <options>
  -action string
        The action (Start or Init). (default "Start")
  -baudrate int
        The baud rate of the serial connection. (default 9600)
  -broker string
        Connect to MQTT broker (e.g. tcp://localhost:1883).
  -device string
        The serial port descriptor. (default "/dev/ttyUSB0")
  -server int
        The server port for the REST service. (default 5701)
  -topic string
        MQTT topic /solar/<topic>/<item>. (default "Growatt")
  -v    Activate verbose logging.

## MQTT

## Status

Up and running with restarts in the morning. As power goes down (sunset) the interface of the inverter will reset, so the init needs to be resend as soon as the inverter comes back to life. This needs polling, as no sign is yet detected which indicates it is powered on again. Note that reading the serial port is blocking without timeout, so additional processes are started to check the communication.

## How to build and run:

It is written in GoLang. For compilation, see the build.sh script.

Binaries for the Raspberry Pi and Odroid are available though.
* growatt_pi2 : Older B models, Zero possibly
* growatt_pi : Model 3(+)
* growatt_odroid : C2 (maybe others)

Run as ```./growatt --device /dev/ttyUSB0 --baudrate 9600``` or without any arguments to use the default shown.
Currently, stop bits, parity, etc. are fixed.

If you want to initialise the inverter manually, use ```./growatt --action Init```.
 
## Required

You need a 'USB to serial' converter. Remove the little plate to expose the RS2323 port and connect the cable. Connect the USB-side to a Raspberry Pi or other device. Using a Raspberry the serial output should NOT be activated (raspi-config).

Example cable: https://www.aliexpress.com/item/Adapter-Convertor-USB-to-RS232-Serial-Port-9-Pin-DB9-Cable-Serial-COM-Port-free/32318476207.html

## Openhab HTTP

If you would like to use this as Openhab growatt publisher, use this in combination with the HTTP binding: https://www.openhab.org/addons/bindings/http1/

In your ```services/http.cfg```, add the following:
```
growatt.url=http://127.0.0.1:5701/status
growatt.updateInterval=5000
```

Create an ```items/growatt.items``` with the following content:
```
Number Growatt_PvVoltage       "PV voltage [%.1f V]"           { http="<[growatt:5000:JSONPATH($.VoltagePV1)]" }
Number Growatt_GridVoltage     "Net voltage [%.1f V]"          { http="<[growatt:5000:JSONPATH($.VoltageGrid)]" }
Number Growatt_TotalProduction "Production overall [%.1f kWh]" { http="<[growatt:5000:JSONPATH($.TotalProduction)]" }
Number Growatt_DayProduction   "Production today: [%.1f kWh]"  { http="<[growatt:5000:JSONPATH($.DayProduction)]" }
Number Growatt_Frequency       "Net frequency [%.1f Hz]"       { http="<[growatt:5000:JSONPATH($.Frequency)]" }
Number Growatt_Power           "Current power [%.1f W]"        { http="<[growatt:5000:JSONPATH($.Power)]" }
String Growatt_Status          "Status [%s]"                   { http="<[growatt:5000:JSONPATH($.Status)]" }
String Growatt_Update          "Updated: [%s]"                 { http="<[growatt:5000:JSONPATH($.Timestamp)]" }
```

Install the JsonPath transformation plugin.

Note that on power down of the inverter there will be values missing from the JSON, causing messages in the openhab log. As far as I know there is no way to indicate a field is optional for JSONPATH.

## Openhab MQTT

* Install a MQTT broker somewhere (e.g. Mosquitto)
* Install the MQTT binding
* Install a Thing of type MQTT broker and configure for your server
* Install a Thing of type Generic MQTT Thing and select the broker
  (note: this application does not support login to the broker)
* Monitor if you receive messages (e.g. `mosquitto_sub -v -t '#'`)
* Add channels and items to the Generic Thing
* Use as MQTT channels (the 'Growatt' part is configurable):
```
	/solar/Growatt/Power         
	/solar/Growatt/VoltagePV1    
	/solar/Growatt/VoltagePV2    
	/solar/Growatt/VoltageBus    
	/solar/Growatt/VoltageGrid   
	/solar/Growatt/TotalProduction
	/solar/Growatt/DayProduction 
	/solar/Growatt/Frequency     
	/solar/Growatt/Temperature   
	/solar/Growatt/OperationHours
	/solar/Growatt/Status        
	/solar/Growatt/FaultCode     
	/solar/Growatt/Timestamp
```

## Disclaimer

No, this isn't production ready quality code. See License.

The init string for the converter is fixed. 

## License:

*MIT*

Copyright 2018-2021 lemval@gmail.com

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


release/ReleaseNotes.md

# Version 1.2 @ 11.08.2021

Fix: If using MQTT, keep on working when connection drops
Add: Allow compilation for golang 1.16 (using build.sh)


main.go
// Growatt project main.go
package main

import (
	"flag"
	"fmt"

	"os"
	"strings"

	"growattrr/diag"
	"growattrr/reader"
)

var speed int
var port int
var device string
var action string
var broker string
var topic string
var verbose bool

func init() {
	flag.StringVar(&action, "action", "Start", "The action (Start or Init).")
	flag.StringVar(&device, "device", "/dev/ttyUSB0", "The serial port descriptor.")
	flag.StringVar(&broker, "broker", "", "Connect to MQTT broker (e.g. tcp://localhost:1883).")
	flag.StringVar(&topic, "topic", "Growatt", "MQTT topic /solar/<topic>/<item>.")
	flag.IntVar(&speed, "baudrate", 9600, "The baud rate of the serial connection.")
	flag.IntVar(&port, "server", 5701, "The server port for the REST service.")
	flag.BoolVar(&verbose, "v", false, "Activate verbose logging.")
}

func main() {
	diag.Info("Starting Growatt Inverter Reader v1.200")

	//	Read the command line arguments
	flag.Parse()

	diag.Verbosive = verbose

	// Initialize the reader
	serialReader := reader.NewReader(device, speed)

	//	Handle the 'init' command to to send the message to
	//	start the logging of the data.

	if strings.Compare("Init", action) == 0 {
		actionInit(serialReader)
	} else if strings.Compare("Start", action) == 0 {
		actionStart(serialReader)
	} else {
		fmt.Printf("\n == ERROR ==============================")
		fmt.Printf("\n    Invalid action '%s'!", action)
		fmt.Printf("\n =======================================")
		fmt.Printf("\n Usage: %s [<options>]", os.Args[0])
		flag.PrintDefaults()
	}
}

func actionInit(reader *reader.Reader) {
	diag.Info("Init requested...")
	status := reader.InitLogger()
	if status {
		diag.Info("Sent. Please restart!")
	} else {
		diag.Warn("Failed. Please retry!")
	}
}

func actionStart(reader *reader.Reader) {

	// Initialize the interpreter and publisher and start all threads to
	// read data, interpret to datagrams and publish as json

	interpreter := NewInterpreter(reader.GetQueue())
	publisher := NewPublisher()

	go reader.StartMonitored()
	go interpreter.start()
	go publisher.start(port)

	publisher.listen(interpreter, reader)
}
